app/services/action_builder.rb

```diff
@@ -20,6 +20,15 @@ def build_action(extra_attrs = {})
   def previous_action
```

Comment:
Hey @pascalbetz - I'm a little confused by your first two comments. in both the first lookup and the lookup in the loop, I'm using `||=`. That memoizes it, so if `@previous_action` is already set, the lookup is skipped. Am I missing something? That use of `||=` is key for this entire method. 

Good note with the find_by.